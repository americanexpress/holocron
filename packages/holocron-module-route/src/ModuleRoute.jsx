/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/prefer-stateless-function --  extend component */
import { Component } from 'react';
import PropTypes from 'prop-types';
import { createModuleRouteFromElement } from './ModuleRouteUtils';

class ModuleRoute extends Component {
  render() {
    console.error(
      '<ModuleRoute> elements are for router configuration only and should not be rendered'
    );
    return null;
  }
}

ModuleRoute.createRouteFromReactElement = createModuleRouteFromElement;
ModuleRoute.propTypes = {
  // these are not used directly, but are used by createModuleRouteFromElement
  /* eslint-disable react/no-unused-prop-types -- comment above */
  moduleName: PropTypes.string,
  path: PropTypes.string,
  title: PropTypes.string,
  hasChildRoutes: PropTypes.bool,
  /* eslint-enable react/no-unused-prop-types -- enable rule */
};
ModuleRoute.defaultProps = {
  moduleName: undefined,
  path: undefined,
  title: undefined,
  hasChildRoutes: false,
};
ModuleRoute.displayName = 'ModuleRoute';

export default ModuleRoute;
/* eslint-enable react/prefer-stateless-function --  extend component */
