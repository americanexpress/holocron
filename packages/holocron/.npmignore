__tests__
*.tgz
.babelrc