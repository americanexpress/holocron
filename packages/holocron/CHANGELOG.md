# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.1.2](https://github.com/americanexpress/holocron/compare/v1.1.1...v1.1.2) (2020-10-01)


### Bug Fixes

* **holocronModule:** remove check for initial-state ([#56](https://github.com/americanexpress/holocron/issues/56)) ([76c73b1](https://github.com/americanexpress/holocron/commit/76c73b175802ed199ee74ae8587b30c313fda592))





# [1.1.1](https://github.com/americanexpress/holocron/compare/v1.1.0...v1.1.1) (2020-09-15)


### Bug Fixes

* **holocron:** ignore block list in development ([c21f5a4](https://github.com/americanexpress/holocron/commit/c21f5a4ce540255be01d002b0fddee94b4e06a14))
* **holocron:** initial state check ([af88adb](https://github.com/americanexpress/holocron/commit/af88adbc61251e1446a62e6cca212250a166e945))
* **holocron/load:** prevent rebuilding without reducer ([9fd4161](https://github.com/americanexpress/holocron/commit/9fd4161c26235171bbac910ee5aaeb82b9a62fa8))


# [1.1.0](https://github.com/americanexpress/holocron/compare/v1.0.0...v1.1.0) (2020-03-27)


### Bug Fixes

* **vulnerability/lerna:** update to fix security vulnerability ([#3](https://github.com/americanexpress/holocron/issues/3)) ([570a0b8](https://github.com/americanexpress/holocron/commit/570a0b80885ac67b0a2a5e039913f7bd53f16afb))


### Features

* **holocron-config:** added holocron module config ([af45613](https://github.com/americanexpress/holocron/commit/af4561392d220f7dce25f6c5f577cae85a7ad3ed))
* **loadModule:** optional cache bust key ([#11](https://github.com/americanexpress/holocron/issues/11)) ([bf1bc27](https://github.com/americanexpress/holocron/commit/bf1bc277bf571497818505f09073528941b5f868))
* **loadModule:** rename "key" to "clientCacheRevision" ([#20](https://github.com/americanexpress/holocron/issues/20)) ([1880972](https://github.com/americanexpress/holocron/commit/188097210b9722df4fa02d2081cb004367d53387))
