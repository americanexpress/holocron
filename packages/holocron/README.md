<h1 align="center">
  <img src='https://github.com/americanexpress/holocron/raw/main/holocron.png' alt="Holocron - One Amex" width='50%'/>
</h1>

[![npm](https://img.shields.io/npm/v/holocron)](https://www.npmjs.com/package/holocron)

> Use this to compose and load your application modules.

## 📖 Table of Contents

* [Features](#-features)
* [API](#%EF%B8%8F-api)
* [Available Scripts](#-available-scripts)

## ✨ Features

### How it works

#### Module Registry

Holocron maintains an in-memory registry of Modules that can be updated dynamically without
requiring a server restart. The idea is that an application can update Holocron's module registry
whenever a new Holocron Module is to be added to the application's runtime. The end result is an
application that can have React components updated/added to it at runtime.

#### Holocron store

After the server is initialized the holocron store is created within `createHolocronStore`, this
extends the Redux store and adds the reducers of your module to the store.

#### Compose modules

Once the modules to render are retrieved we dispatch  `composeModules` to get the data for these
modules, this data is needed for rendering.

## 🎛️ API

Check out the [API reference guide](./docs/api/README.md).

## 📜 Available Scripts

To test out any changes that you've made locally, run `yarn pack` then install this within your
application.

The scripts below are available to run and use:

**`yarn prebuild`**

This removes any existing files generated during the build process and ensures that any new build is
clean.

**`yarn build`**

This deletes the current generated JS files within the directory and compiles the ECMAScript 6 code
within the `src` file to a version of ECMAScript that can run in current browsers using Babel
afterwards it copies them to the lib folder.

**`yarn prepublish`**

This runs `yarn build`

**`yarn watch:build`**

This watches for any changes in the  `src` folder and runs `yarn build` if any changes are
detected.
