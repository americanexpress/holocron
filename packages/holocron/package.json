{
  "name": "holocron",
  "version": "1.1.2",
  "main": "index.js",
  "module": "src/index.js",
  "keywords": [
    "holocron",
    "amex",
    "one-app"
  ],
  "description": "Use this to compose and load your application modules.",
  "contributors": [
    "Andres Escobar <Andres.Escobar@aexp.com> (https://github.com/anescobar1991)",
    "James Singleton <James.Singleton1@aexp.com> (https://github.com/JamesSingleton)",
    "Jimmy King  <Jimmy.King@aexp.com> (https://github.com/10xLaCroixDrinker)",
    "Jonathan Adshead <Jonathan.Adshead@aexp.com> (https://github.com/JAdshead)",
    "Michael Tobia <Michael.M.Tobia@aexp.com> (https://github.com/Francois-Esquire)",
    "Michael Tomcal <Michael.A.Tomcal@aexp.com> (https://github.com/mtomcal)",
    "Stephanie Coates  <Stephanie.Coates1@aexp.com> (https://github.com/stephaniecoates)",
    "Nelly Kiboi <Nelly.J.Kiboi@aexp.com> (https://github.com/nellyk)",
    "Nickolas Oliver <nickolas.oliver@aexp.com> (https://github.com/PixnBits)"
  ],
  "scripts": {
    "clean": "rimraf *.js server ducks",
    "build": "babel src -d ./",
    "prebuild": "yarn clean",
    "prepack": "yarn build",
    "prepublish": "yarn build",
    "watch:build": "nodemon --watch src --ext js,jsx --exec yarn build"
  },
  "license": "Apache-2.0",
  "dependencies": {
    "@americanexpress/vitruvius": "^2.0.0",
    "hoist-non-react-statics": "^3.3.0",
    "immutable": "^4.0.0-rc.12",
    "prop-types": "^15.5.10",
    "react": "^16.11.0",
    "react-dom": "^16.11.0",
    "react-redux": "^7.1.1",
    "redux": "^4.0.4",
    "redux-immutable": "^4.0.0",
    "redux-thunk": "^2.1.0",
    "require-from-string": "^1.2.1",
    "reselect": "^4.0.0",
    "shallowequal": "^1.1.0",
    "ssri": "^7.0.1"
  },
  "devDependencies": {
    "@babel/polyfill": "^7.2.5",
    "nodemon": "^1.10.2",
    "rimraf": "^2.5.4"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/americanexpress/holocron.git",
    "directory": "packages/holocron"
  },
  "homepage": "https://github.com/americanexpress/holocron",
  "bugs": {
    "url": "https://github.com/americanexpress/holocron/issues"
  }
}
