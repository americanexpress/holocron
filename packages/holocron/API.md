# 🎛 API

There are some functions that are available on both the client and server, and some that are only
available on the server. We have organized them as such.

- [Universal](#universal)
- [Server](#server)

## Universal

**Contents:**

- [App-level Functions](#app-level-functions)
  - [`createHolocronStore`](#createholocronstore)
- [Holocron Module Configuration](#holocron-module-configuration)
  - [`Module.holocron`](#holocron)
- [Module-level Functions](#module-level-functions)
  - [`holocronModule`](#holocronmodule)
  - [`RenderModule`](#rendermodule)
  - [`composeModules`](#composemodules)
  - [`loadModule`](#loadmodule)
- [Module Registry](#module-registry)
  - [`registerModule`](#registermodule)
  - [`getModule`](#getmodule)
  - [`getModules`](#getmodules)
  - [`getModuleMap`](#getmodulemap)
  - [`setModuleMap`](#setmodulemap)
- [Selectors](#selectors)
  - [`isLoaded`](#isloaded)
  - [`failedToLoad`](#failedToLoad)
  - [`getLoadError`](#getLoadError)
  - [`isLoading`](#isloading)
  - [`getLoadingPromise`](#getloadingpromise)

### App-level Functions

The below is intended only to be called by your app, not modules.

#### `createHolocronStore`

Creates the [Redux] store with Holocron compatibility.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `reducer` | `(state, action) => newState` | `true` | The [Redux reducer] for your application |
| `initialState` | `Immutable.Map` | `false` | The initial state of your [Redux] store |
| `enhancer` | `Function` | `false` | A [Redux enhancer] |
| `localsForBuildInitialState` | `Object` | `false` | Value to pass to [vitruvius]'s `buildInitialState` |
| `extraThunkArguments` | `Object` | `false` | Additional arguments to be passed to [Redux thunks] |

##### Usage

```jsx
import { createHolocronStore } from 'holocron';
import { Provider } from 'react-redux';

const store = createHolocronStore({
  reducer,
  initialState,
  enhancer,
  localsForBuildInitialState,
  extraThunkArguments,
});

hydrate(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
);
```

### Holocron Module Configuration

#### `Module.holocron`

The optional `holocron` object set to the parent React Component inside a Holocron Module determines behavior of state management, data loading, and React prop management. 

> A `name` property is required if a `reducer` is set otherwise the `reducer` will not be added to the Redux Store.

##### Properties

| name | type | required | value |
|---|---|---|---|
| `name` | `String` | `true` | The name of your Holocron module |
| `reducer` | `(state, action) => newState` | `false` | The Redux reducer to register when your module is loaded. *Requires a `name`* |
| `load` *☠️ Deprecated* | `(props) => Promise` or `(props) => (dispatch, getState, ...extra) => Promise` | `false` | A deprecated function that fetches data required by your module. Please use `loadModuleData` instead. |
| `loadModuleData` | `({ store, fetchClient, ownProps, module }) => Promise` | `false` | A function that fetches data required by your module |
| `shouldModuleReload` | `(oldProps, newProps) => Boolean` | `false` | A function to determine if your `loadModuleData` and or `load` function should be called again |
| `mergeProps` | `(stateProps, dispatchProps, ownProps) => Object` | `false` | Passed down to Redux connect |
| `options` | `Object` | `false` | Additional options |

#### Usage

```jsx
import React from 'react';
import PropTypes from 'prop-types';
import { reducer, fetchData } from '../duck';

const HelloWorld = ({ moduleState: { myData } }) => (
  <h1>
Hello,
    {myData.name}
  </h1>
);
HelloWorld.propTypes = {
  moduleState: PropTypes.shape({
    myData: PropTypes.string,
  }).isRequired,
};

const loadModuleData = ({ store: { dispatch }, ownProps }) => dispatch(fetchData(ownProps.input));
const shouldModuleReload = (oldProps, newProps) => oldProps.input !== newProps.input;

HelloWorld.holocron = {
  name: 'hello-world',
  reducer,
  loadModuleData,
  shouldModuleReload,
  options,
}

export default HelloWorld;
```

The Holocron Module parent React Components will be provided several props automatically.

| prop name | type | value |
|---|---|---|
| `moduleLoadStatus` | `String` | One of `"loading"`, `"loaded"`, or `"error"`, based on the `load` function |
| `moduleState` | `Object` | The state of the registered reducer after [`.toJS()`] has been called on it |

### Module-level Functions

While these can all be used by the app itself, they will get the most use from modules.

#### `holocronModule`

> ☠️ Deprecated in favor of [Holocron Module Configuration](#holocron-module-configuration)

A [higher order component (HOC)] for registering a load function and/or reducer with a module. This HOC is only required if the `load` or `reducer` functionality is used.

> A `name` property is required if a `reducer` is set otherwise the `reducer` will not be added to the Redux Store.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `name` | `String` | `true` | The name of your Holocron module |
| `reducer` | `(state, action) => newState` | `false` | The Redux reducer to register when your module is loaded. *Requires a `name`* |
| `load` *☠️ Deprecated* | `(props) => Promise` or `(props) => (dispatch, getState, ...extra) => Promise` | `false` | A deprecated function that fetches data required by your module. Please use `loadModuleData` instead. |
| `loadModuleData` | `({ store, fetchClient, ownProps, module }) => Promise` | `false` | A function that fetches data required by your module |
| `shouldModuleReload` | `(oldProps, newProps) => Boolean` | `false` | A function to determine if your `loadModuleData` and or `load` function should be called again |
| `mergeProps` | `(stateProps, dispatchProps, ownProps) => Object` | `false` | Passed down to Redux connect |
| `options` | `Object` | `false` | Additional options |

##### Usage

```jsx
import { holocronModule } from 'holocron';
import React from 'react';
import PropTypes from 'prop-types';
import { reducer, fetchData } from '../duck';

const HelloWorld = ({ moduleState: { myData } }) => (
  <h1>
Hello,
    {myData.name}
  </h1>
);
HelloWorld.propTypes = {
  moduleState: PropTypes.shape({
    myData: PropTypes.string,
  }).isRequired,
};

const load = (props) => (dispatch) => dispatch(fetchData(props.input));
const shouldModuleReload = (oldProps, newProps) => oldProps.input !== newProps.input;
const options = { ssr: true };

export default holocronModule({
  name: 'hello-world',
  reducer,
  load,
  shouldModuleReload,
  options,
})(HelloWorld);
```

Components using this HOC will be provided several props.

| prop name | type | value |
|---|---|---|
| `moduleLoadStatus` | `String` | One of `"loading"`, `"loaded"`, or `"error"`, based on the `load` function |
| `moduleState` | `Object` | The state of the registered reducer after [`.toJS()`] has been called on it |

#### `RenderModule`

A React component for rendering a Holocron module.

##### Props

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `PropTypes.string` | `true` | The name of the Holocron module to be rendered |
| `props` | `PropTypes.object` | `false` | Props to pass the rendered Holocron module |
| `children` | `PropTypes.node` | `false` | Childen passed to the rendered Holocron module |

##### Usage

```jsx
import { RenderModule, holocronModule, composeModules } from 'holocron';

const MyComponent = ({ data }) => (
  <div>
    {/* some more JSX */}
    <RenderModule moduleName="sub-module" props={{ data }}>
      <p>Hello, world</p>
    </RenderModule>
  </div>
);

const load = ({ data }) => (dispatch) => dispatch(composeModules([{ name: 'sub-module', props: { data } }]));

export default holocronModule({
  name: 'my-module',
  load,
})(MyComponent);
```

#### `composeModules`

An action creator that loads Holocron modules and their data.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleConfigs` | `[{ name, props }]` | `true` | An array of objects containing module names and their props |

##### Usage

```js
import { composeModules } from 'holocron';

const load = (props) => (dispatch) => dispatch(composeModules([
  { name: 'some-module', props: { someProp: props.anyProp } },
  { name: 'another-module' },
]));
```

#### `loadModule`

An action creator that fetches a Holocron module.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module being fetched |

##### Usage

```js
import { loadModule } from 'holocron';

const load = () => (dispatch) => dispatch(loadModule('my-module'));
```

> The following are low-level APIs unlikely to be needed by most users

### Module Registry

These functions are all related to interactions with the module registry.

#### `registerModule`

Adds a Holocron module to the registry

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of your Holocron module |
| `module` | `Function` | `true` | The Holocron module itself (a React component) |

#### `getModule`

Retrives a Holocron module from the registry

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module being requested |
| `altModules` | `Immutable.Map` | `false` | An alternative set of modules to the registry |

##### Usage

```js
import { getModule } from 'holocron';

const Module = getModule(moduleName, altModules);
```

#### `getModules`

Returns all modules in the registry

##### Arguments

_none_

##### Usage

```js
import { getModules } from 'holocron';

const modules = getModules();
```

#### `getModuleMap`

Returns the module map

##### Arguments

_none_

##### Usage

```js
import { getModuleMap } from 'holocron';

const moduleMap = getModuleMap();
```

#### `setModuleMap`

Sets the module map

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `newModuleMap` | `Immutable.Map` | `true` | The new module map to replace the existing one |

##### Usage

```js
import { setModuleMap } from 'holocron';

setModuleMap(newModuleMap);
```

### Selectors

These are all functions that take a module name as the only parameter and return a new function
that accepts the [Redux] state as the only parameter which returns data about the module.

#### `isLoaded`

A selector to determine if a Holocron module has been loaded.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module that may be loaded |

##### Usage

```js
import { isLoaded } from 'holocron';

const mapStateToProps = (state) => (
  { myModuleIsLoaded: isLoaded('my-module')(state) }
);
```

#### `failedToLoad`

A selector to determine if a Holocron module failed to load.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module that may have failed to load |

##### Usage

```js
import { failedToLoad } from 'holocron';

const mapStateToProps = (state) => ({
  myModuleFailedToLoad: failedToLoad('my-module')(state),
});
```

#### `getLoadError`

A selector to return the error of a Holocron module that failed to load.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module whose load error will be returned |

##### Usage

```js
import { getLoadError } from 'holocron';

const mapStateToProps = (state) => ({
  myModuleLoadError: getLoadError('my-module')(state),
});
```

#### `isLoading`

A selector to determine if a module is loading.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module that may be loading |

##### Usage

```js
import { isLoading } from 'holocron';

const mapStateToProps = (state) => ({
  myModuleIsLoading: isLoading('my-module')(state),
});
```

#### `getLoadingPromise`

A selector to return the promise from a Holocron module being loaded.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleName` | `String` | `true` | The name of the Holocron module whose loading promise will be returned |

##### Usage

```js
import { getLoadingPromise } from 'holocron';

const mapStateToProps = (state) => ({
  myModuleLoadingPromise: getLoadingPromise('my-module')(state),
});
```

## Server

**Contents:**

- [`updateModuleRegistry`](#updatemoduleregistry)
- [`areModuleEntriesEqual`](#aremoduleentriesequal)

#### `updateModuleRegistry`

Updates the module registry with a new module map.

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `moduleMap` | `Object` | `true` | The new module map |
| `onModuleLoad` | `Function` | `false` | The function to call on every module that is loaded |
| `batchModulesToUpdate` | `modules => Array` | `false` | A function that returns an array of arrays of batches of modules to load |
| `getModulesToUpdate` | `Function` | `false` | A function that returns an array of which modules should be updated |


##### Usage

```js
import { updateModuleRegistry } from 'holocron/server';

const onModuleLoad = ({ module, moduleName }) => {
  if (!isValid(module)) throw new Error(`Module ${moduleName} is invalid`);
  console.info(`Loaded module ${moduleName}`);
};

export default async function() {
  const moduleMapResponse = await fetch(MODULE_MAP_URL);
  const moduleMap = await moduleMapResponse.json();
  await updateModuleRegistry({
    moduleMap,
    onModuleLoad,
  });
}

```

#### `areModuleEntriesEqual`

Compares two module map entries to see if they are equal. This is intended for use when providing `getModulesToUpdate` to `updateModulesRegistry`

##### Arguments

| name | type | required | value |
|---|---|---|---|
| `firstModuleEntry` | `Object` | `false` | A module map entry |
| `secondModuleEntry` | `Object` | `false` | Another module map entry |

##### Usage

```js
import { areModuleEntriesEqual } from 'holocron/server';

const getModulesToUpdate = (
  currentModules, nextModules) => Object.keys(next).filter((moduleName) => (
  !areModuleEntriesEqual(curr[moduleName], next[moduleName])
        || someOtherLogic(moduleName))
);
```

[Redux]: https://redux.js.org
[Redux enhancer]: https://redux.js.org/recipes/configuring-your-store#extending-redux-functionality
[Redux reducer]: https://redux.js.org/recipes/structuring-reducers/structuring-reducers
[vitruvius]: http://github.com/americanexpress/vitruvius
[Redux thunks]: https://github.com/reduxjs/redux-thunk
[`.toJS()`]: https://immutable-js.github.io/immutable-js/docs/#/Map/toJS
[higher order component (HOC)]: https://reactjs.org/docs/higher-order-components.html
