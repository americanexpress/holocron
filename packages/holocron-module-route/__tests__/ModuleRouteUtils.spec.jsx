/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
// holocron is a peer dependency, I would add it to devDeps, but it is immediately mocked
import { loadModule } from 'holocron'; // eslint-disable-line import/no-unresolved,import/extensions
import ModuleRoute from '../src/ModuleRoute';
import {
  addToRouteProps,
  passChildrenProps,
  getRouteIndex,
  createModuleRoute,
  createModuleRouteFromElement,
} from '../src/ModuleRouteUtils';

jest.mock('holocron', () => ({
  loadModule: jest.fn((moduleName) => {
    const modules = {
      'test-module': {
        childRoutes: [{
          moduleName: 'another-module',
          path: 'another',
          indexRoute: 'index',
        }],
      },
      'store-hook-module': {
        onEnterRouteHook: (store) => (nextState, replace) => {
          store.dispatch({ type: 'NEXT_STATE', nextState });
          replace('/test-store-hook');
        },
      },
      'sync-hook-module': {
        onEnterRouteHook: jest.fn((nextState, replace) => replace('/test-sync-hook')),
      },
      'async-hook-module': {
        onEnterRouteHook: (nextState, replace, cb) => {
          replace('/test-async-hook');
          cb('async hook done');
        },
      },
    };

    return Promise.resolve(modules[moduleName]);
  }),
}));

describe('ModuleRouteUtils', () => {
  describe('addToRouteProps', () => {
    it('should add props and replace old props with new ones', () => {
      const route = {
        unchanged: 'value',
        props: {
          unchanged: 'value',
          changed: 'old',
        },
      };
      const newProps = {
        changed: 'new',
        new: 'prop',
      };
      expect(addToRouteProps(route, newProps)).toMatchSnapshot();
    });
  });

  describe('passChildrenProps', () => {
    it('should pass props to children', () => {
      const routes = [{}];
      const newProps = { hello: 'world' };
      expect(passChildrenProps(routes, newProps)).toMatchSnapshot();
    });

    it('should convert a single route to an array', () => {
      const routes = passChildrenProps({}, {});
      expect(Array.isArray(routes)).toBe(true);
    });

    it('should pass the store to a function that creates routes', () => {
      const getState = () => 'Hello world';
      const createRoutes = (store) => ({ props: { state: store.getState() } });
      const newProps = { store: { getState } };
      expect(passChildrenProps(createRoutes, newProps)).toMatchSnapshot();
    });

    it('should return an empty array if no routes are provided', () => {
      expect(passChildrenProps(undefined, {})).toMatchSnapshot();
    });
  });

  describe('getRouteIndex', () => {
    it('should return the index route', () => {
      const indexRoute = 'indexRoute';
      const routes = [{ indexRoute }, { indexRoute: 'lol' }];
      expect(getRouteIndex(routes)).toBe(indexRoute);
    });
  });

  describe('createModuleRoute', () => {
    const dispatch = jest.fn((action) => {
      if (typeof action === 'function') {
        return action(dispatch);
      }
      return action;
    });
    const store = { dispatch };
    const defaultProps = {
      default: 'value',
      hello: 'world',
    };

    beforeEach(() => dispatch.mockClear());

    it('should create a valid module route', () => {
      const props = { title: 'Test', moduleName: 'test-module' };
      const moduleRoute = createModuleRoute(defaultProps, props);
      expect(moduleRoute).toMatchSnapshot();
    });

    it('should use the path to generate a title if none exists', () => {
      const props = { path: 'edit', moduleName: 'edit-module' };
      const moduleRoute = createModuleRoute(defaultProps, props);
      expect(moduleRoute.title).toBe('Edit');
    });

    it('should not add moduleRoute props if no moduleName is present', () => {
      const moduleRoute = createModuleRoute(defaultProps, { title: 'test' });
      expect(moduleRoute).toMatchSnapshot();
    });

    describe('getIndexRoute', () => {
      it('should get the index route', (done) => {
        const moduleName = 'test-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);
        const callback = jest.fn(() => {
          expect(loadModule).toHaveBeenCalledWith(moduleName);
          expect(callback.mock.calls[0][1]).toBe('index');
          done();
        });
        return moduleRoute.getIndexRoute(undefined, callback);
      });
    });

    describe('getChildRoutes', () => {
      it('should get the child routes', (done) => {
        const moduleName = 'test-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);
        const callback = jest.fn(() => {
          expect(loadModule).toHaveBeenCalledWith(moduleName);
          expect(callback.mock.calls[0][1]).toMatchSnapshot();
          done();
        });
        return moduleRoute.getChildRoutes(undefined, callback);
      });
    });

    describe('getComponent', () => {
      it('should get the module', (done) => {
        const moduleName = 'test-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);
        const callback = jest.fn(() => {
          expect(loadModule).toHaveBeenCalledWith(moduleName);
          expect(callback.mock.calls[0][1]).toMatchSnapshot();
          done();
        });
        return moduleRoute.getComponent(undefined, callback);
      });
    });

    describe('onEnter', () => {
      it('should allow the Module to specify an onEnter hook that takes the store and returns a hook', (done) => {
        const moduleName = 'store-hook-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);

        const replace = jest.fn();
        function cb(cbMessage) {
          expect(cbMessage).toBeUndefined();
          expect(store.dispatch).toHaveBeenCalledWith({ type: 'NEXT_STATE', nextState: 'nextState' });
          expect(replace).toHaveBeenCalledWith('/test-store-hook');
          done();
        }
        moduleRoute.onEnter('nextState', replace, cb);
      });

      it('should allow the Module to specify a synchronous onEnter hook', (done) => {
        const moduleName = 'sync-hook-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);

        const replace = jest.fn();
        function cb(cbMessage) {
          expect(cbMessage).toBeUndefined();
          expect(replace).toHaveBeenCalledWith('/test-sync-hook');
          done();
        }
        moduleRoute.onEnter('nextState', replace, cb);
      });

      it('should allow the Module to specify an asyncronous hook', (done) => {
        const moduleName = 'async-hook-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);

        const replace = jest.fn();
        function cb(cbMessage) {
          expect(cbMessage).toBe('async hook done');
          expect(replace).toHaveBeenCalledWith('/test-async-hook');
          done();
        }
        moduleRoute.onEnter('nextState', replace, cb);
      });

      it('should allow the Module to not specify an onEnter hook', (done) => {
        const moduleName = 'test-module';
        const props = { moduleName, store };
        const moduleRoute = createModuleRoute(defaultProps, props);

        const replace = jest.fn();
        function cb() {
          // Don't need to assert anything, as long as it reached here everything's good
          done();
        }
        moduleRoute.onEnter('nextState', replace, cb);
      });

      it('should only use the onEnter hook defined by the Module if there is not one defined directly on the route', () => {
        const moduleName = 'sync-hook-module';
        const onEnter = jest.fn();
        const props = { moduleName, store, onEnter };
        const moduleRoute = createModuleRoute(defaultProps, props);

        moduleRoute.onEnter('nextState', 'replace', 'cb');
        expect(onEnter).toHaveBeenCalledWith('nextState', 'replace', 'cb');
      });
    });
  });

  describe('createModuleRouteFromElement', () => {
    it('should convert a jsx ModuleRoute to a valid module route', () => {
      const route = (
        <ModuleRoute path="/" moduleName="parent-module">
          <ModuleRoute path="test" moduleName="test-module" />
        </ModuleRoute>
      );
      const moduleRoute = createModuleRouteFromElement(route);
      expect(moduleRoute).toMatchSnapshot();
    });

    it('should not add childRoutes if the only children are not valid elements', () => {
      const route = (
        // eslint-disable-next-line react/self-closing-comp
        <ModuleRoute path="/" moduleName="parent-module">
          {[]}
        </ModuleRoute>
      );
      const moduleRoute = createModuleRouteFromElement(route);
      expect(moduleRoute).toMatchSnapshot();
    });
  });
});
