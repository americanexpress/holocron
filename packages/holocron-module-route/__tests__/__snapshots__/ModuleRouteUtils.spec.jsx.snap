// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleRouteUtils addToRouteProps should add props and replace old props with new ones 1`] = `
{
  "props": {
    "changed": "new",
    "new": "prop",
    "unchanged": "value",
  },
  "unchanged": "value",
}
`;

exports[`ModuleRouteUtils createModuleRoute getChildRoutes should get the child routes 1`] = `
[
  {
    "indexRoute": "index",
    "moduleName": "another-module",
    "path": "another",
    "props": {
      "store": {
        "dispatch": [MockFunction] {
          "calls": [
            [
              Promise {},
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": Promise {},
            },
          ],
        },
      },
    },
  },
]
`;

exports[`ModuleRouteUtils createModuleRoute getComponent should get the module 1`] = `
{
  "childRoutes": [
    {
      "indexRoute": "index",
      "moduleName": "another-module",
      "path": "another",
    },
  ],
}
`;

exports[`ModuleRouteUtils createModuleRoute should create a valid module route 1`] = `
{
  "default": "value",
  "getChildRoutes": [Function],
  "getComponent": [Function],
  "getIndexRoute": [Function],
  "hello": "world",
  "moduleName": "test-module",
  "onEnter": [Function],
  "title": "Test",
}
`;

exports[`ModuleRouteUtils createModuleRoute should not add moduleRoute props if no moduleName is present 1`] = `
{
  "default": "value",
  "hello": "world",
  "title": "test",
}
`;

exports[`ModuleRouteUtils createModuleRouteFromElement should convert a jsx ModuleRoute to a valid module route 1`] = `
{
  "childRoutes": [
    {
      "getChildRoutes": [Function],
      "getComponent": [Function],
      "getIndexRoute": [Function],
      "hasChildRoutes": false,
      "moduleName": "test-module",
      "onEnter": [Function],
      "path": "test",
      "store": undefined,
      "title": "Test",
    },
  ],
  "getChildRoutes": [Function],
  "getComponent": [Function],
  "getIndexRoute": [Function],
  "hasChildRoutes": false,
  "moduleName": "parent-module",
  "onEnter": [Function],
  "path": "/",
  "title": "/",
}
`;

exports[`ModuleRouteUtils createModuleRouteFromElement should not add childRoutes if the only children are not valid elements 1`] = `
{
  "getChildRoutes": [Function],
  "getComponent": [Function],
  "getIndexRoute": [Function],
  "hasChildRoutes": false,
  "moduleName": "parent-module",
  "onEnter": [Function],
  "path": "/",
  "title": "/",
}
`;

exports[`ModuleRouteUtils passChildrenProps should pass props to children 1`] = `
[
  {
    "props": {
      "hello": "world",
    },
  },
]
`;

exports[`ModuleRouteUtils passChildrenProps should pass the store to a function that creates routes 1`] = `
[
  {
    "props": {
      "state": "Hello world",
      "store": {
        "getState": [Function],
      },
    },
  },
]
`;

exports[`ModuleRouteUtils passChildrenProps should return an empty array if no routes are provided 1`] = `[]`;
